doc/rplctl.8
.TH "rplctl" "8" "2009-09-28" "ttyrpld" "tty logging daemon suite"
.SH "Name"
.PP
rplctl \(em control rpld
.SH Syntax
.PP
\fBrplctl\fP [\fB\-t\fP] [\fB\-A\fP \fItty\fP] [\fB\-D\fP \fItty\fP] [\fB\-L\fP
[\fItty\fP]] [\fB\-S\fP \fItty\fP] [\fB\-X\fP \fItty\fP] [\fB\-Z\fP
[\fItty\fP]] [\fB\-f\fP \fIsocket\fP]
.SH Description
.PP
The \fBrplctl\fP utility is used to query and control rpld (and its infod
subcomponent). It will print info about the requested ttys, or all ttys
currently monitored or optionized if no arguments are given.
.PP
rplctl will read the INFOD_PORT variable from rpld.conf by default, if not
overriden by the \fB\-f\fP option.
.SH Options
.PP
.TP
\fB\-A\fP {\fItty\fP|\fB=\fP\fIMAJOR\fP\fB,\fP\fIMINOR\fP}
Activate logging for the given tty. It can be a filename (which must exist) or
a device number using "=MAJOR,MINOR".
.TP
\fB\-D\fP \fItty\fP
Deactivate logging for the given tty.
.TP
\fB\-L\fP [\fItty\fP]
Explicitly request a listing of all ttys currently active or have options set.
If a tty is given, only display info for that particular one.
.TP
\fB\-S\fP \fItty\fP
Deactivate logging for the given tty.
.TP
\fB\-X\fP \fItty\fP
Close the tty's log file. rpld will open up a new one whenever new activity is
detected (and logging this terminal is not deactivated) on that terminal.
.TP
\fB\-Z\fP [\fItty\fP]
Zero all packet and byte counters, both of rpld and every single tty. If a tty
is given, only zeroes the stats of that tty.
.TP
\fB\-f\fP \fIsocket\fP
Path to the infod socket to connect to. The default is /var/run/.rplinfo_socket
or INFOD_PORT in the rpld.conf configuration file.
.TP
\fB\-t\fP
Generate output suitable for sscanf(), split() or anything that deals best with
simple text strings. See below for details on the output format.
.SH "Example output"
.PP
rplctl without arguments could print:
.PP
.nf
A TTY      BYTES IN     OUT FILENAME
========================================================
  (bytes)      2159   81129 ttyrpld 2.52/Linux
  (packets)      14     250
--------------------------------------------------------
D tty7          937       0 root/20050328.224208.tty7
D pts-1           0     586 boinc/20050328.224219.pts-1
D pts-2           0    5812 root/20050328.224208.pts-2
D pts-3         358   22113 root/20050328.224208.pts-3
D pts-4         864   52618 root/20050328.224341.pts-4
--------------------------------------------------------
.fi
.PP
The status field (A) can be one of \fBA\fP, \fBD\fP or \fBS\fP: Activated,
Deactivated, Deactived for this Session.
.SS "sscanf format"
.PP
On the other hand, `rplctl -t` could produce this:
.PP
.nf
ttyrpld 2.52
format 4
0 144 143 0 3347 2496 3756 122881 0 0
D 4 7 tty7 1507 0 /var/log/rpl/root/20050328.224208.tty7
D 136 1 pts-1 0 1241 /var/log/rpl/boinc/20050328.224219.pts-1
D 136 2 pts-2 0 7585 /var/log/rpl/root/20050328.224208.pts-2
D 136 3 pts-3 365 22798 /var/log/rpl/root/20050328.224208.pts-3
D 136 4 pts-4 1884 91257 /var/log/rpl/root/20050328.224341.pts-4
.fi
.PP
The first line of the "parseable" output (-t) is the program used. It can (and
should) be ignored by programs querying infod. The second is the status line
about rpld. The sscanf formula is "%lu %lu %lu %lu %llu %llu %lu %lu". The
fields are (in this order):
.IP \(bu 4
RPLEVT_OPEN \(em number of open() operation on any tty
.IP \(bu 4
RPLEVT_LCLOSE \(em final tty closes
.IP \(bu 4
RPLEVT_READ \(em payload bytecount
.IP \(bu 4
RPLEVT_WRITE \(em payload bytecount
.IP \(bu 4
RPLEVT_READ \(em packets read from rpldev
.IP \(bu 4
RPLEVT_WRITE \(em packets read from rpldev
.IP \(bu 4
packets considered bad (i.e. bad magic) \(em might be more than there are
actually damaged, because the algorithm needs to find something that looks
good.
.PP
All other lines are tty information lines, whose formula is "%c %ld %ld %s %llu
%llu %s", and the fields are:
.IP \(bu 4
status \(em 'A' for activated, 'D' is deactivated and 'S' means deactivated
until session ends.
.IP \(bu 4
major number
.IP \(bu 4
minor number
.IP \(bu 4
mnemonic string for the major/minor number
.IP \(bu 4
RPLEVT_READ payload bytecount
.IP \(bu 4
RPLEVT_WRITE payload bytecount
.IP \(bu 4
full filename being written to (this is different from the human-readable
output which only shows it in part)
.SH "See also"
.PP
\fBrpld\fP(8)


doc/changelog.txt

Notes
=====

You do not need to repatch your kernel or recompile rpldev if they have not
changed significantly between ttyrpld releases.

│                │last time kpatch│                            │
│Operating system│(rpldhk) changed│last time rpldev changed:   │
├────────────────┼────────────────┼────────────────────────────┤
│Linux           │2.50            │2.60                        │
│FreeBSD         │2.60            │2.60                        │
│OpenBSD         │-               │2.60                        │
│NetBSD          │2.50            │2.60                        │


Changelog
=========

HEAD

Changes:
- doc: move changelog to text/plain
- rpldev: NetBSD 5.0.1 support
- rpldev: OpenBSD 4.5 support
- rpldev: Linux 2.6.31 support
- rpldev: FreeBSD 8.0 support
- packet: increase the "size" field to 32 bits
- packet: combine "magic" and "event" fields, and move to front


v2.52 (2008-09-07)

  • Updated kernel components for Linux 2.6.27
  • Updated userspace code for libHX 1.25

v2.51 (2008-05-15)

  • Updated rpldhk and rpldev for Linux 2.6.25, OpenBSD 4.3, FreeBSD 7.0,
    NetBSD 4.0

v2.50 (2008-01-01)

  • changed the packet format to be compatible between 32-bit and 64-bit
  • added rplcvt utility for converting old on-disk rpld files
  • explicitly blacklist /dev/stdout, /dev/stderr and /dev/stderr - these have
    been popping up as tty names because there is such a symlink in Linux's /
    dev. (But we need symlink processing for Solaris, eww.)

v2.19 (2007-09-17)

  • The kernel interface has been reduced (hooks removed). The 2.19 userspace
    is still compatible (throws warnings though) with 2.18 kernel components
    however. The 2.19 kernel patch is for 2.6.22.
  • an experimental Solaris kernel driver has been added - tested with Solaris
    11.snv only, but may work with older ones.
  • fix a file descriptor leak in rpld (thanks J.Eller)
  • fix a symlink recursion when walking /dev (J.Eller)
  • raise maximum number of open files (J.Eller)
  • handle permission restrictions on /dev nodes
  • fix directory name of Persian translation
  • moved to autotools
  • build with new libHX 1.10

v2.18 (2006-07-01)

  • Fixed OpenBSD 3.9 kpatch
  • Readded workaround for OpenBSD/NetBSD EINTR signal handling that got lost
    in 2.17

v2.17 (2006-06-19)

  • Support for Linux 2.6.17, FreeBSD 6.1, OpenBSD 3.9
  • added missing -fprofile-arcs -ftest-coverage to LDFLAGS for PROF=1

v2.16 (2006-04-01)

  • user/rpld.c: removed a superfluous variable
  • Added polling functions to the BSD rpldevs. This fixes the high CPU usage
    (which is due to the implementation of pthreads on OpenBSD) observed with
    rpld/OpenBSD.
  • OpenBSD part is again just a patch since modules are said to be unsupported
    on non-i386
  • Due to popular request, the docs now come with a black-on-white ("paper
    color") color scheme and no fixed font size. The old gray-on-black
    ("console color") is available as an alternate stylesheet and can be used
    in Firefox using View > Page Style > Black background.
  • Added the EVT_MAGIC pseudo event type
  • Added share/ttyrpld.magic for the file(1) utility

v2.15 (2006-03-11)

General:

  • added support for NetBSD 3.0
  • shifts for majors>255 and minors>255 were broken in include/dev.h

kpatch:

  • [All] fixed some SMP race conditions
  • [OpenBSD] fixed: rpl_deinit() was not called in ttyclose()

rpldev:

  • [All] rename functions to make function more clear (mv_to_user ->
    circular_get, mv_buffer -> circular_put_packet, mv_buffer2 -> circular_put)
  • [All] make avail_R() and avail_W() return size_t instead of ssize_t,
    addressing some potential overflow problems
  • [Linux] document circulary calculation of avail_*(), please report errors
  • [FreeBSD] merged freebsd-*/rpldev.c into one file
  • [FreeBSD/OpenBSD] removed SKIP_PTM, since master ptys do not generate any
    noise (that's a kpatch feature)
  • [OpenBSD] rpldev is now a module (rpldev_mod.o)
  • [OpenBSD] loads at new device number (228,0), please re-mknod your /dev/rpl
    node

v2.12 (2006-01-21)

  • fixed: FreeBSD 6.0 hands krpl_close() a NULL pointer in some cases; handle
    it properly
  • some doc updates
  • GNUmakefile: "all" target does not include "doc" anymore, because BSDs
    (standard installation) often ship without php, w3m and perl.
  • added the '[' and ']' keys to ttyreplay (lower/raise speed by 0.10)

v2.11 (2006-01-06)

Added support back in for the latest releases of the previous generation of
kernels. For Linux, that's 2.4.something (USE DISCOURAGED!), and for FreeBSD
that's 5.3.

v2.10 (2005-12-04)

Support for OpenBSD 3.8 and FreeBSD 6.0 was added; permission denied, syslog
spamming, cpu hog, ownership and symlink recursion bugs were fixed. Docs were
updated to reflect advances since the last seven releases.

  • Support for OpenBSD 3.8 and FreeBSD 6.0
  • Documentation has been updated to reflect advances since the last seven
    releases.
  • user/replay.c: fixed a color bug when -t was active (spotted whilst under
    FreeBSD)
  • user/*.c: use usleep(10000) instead of sched_yield() to reduce idle cpu hog
  • user/rpld.c: umask was set wrong, so 1.) logging did not take place 2.)
    syslog was spammed with permission denied warnings
  • user/rpld.c: added a rate limiter for syslog messages
  • user/rpld.c: find_devnode_dive() forgot to check for symlinks so that we
    dived into /dev/fd, which is not wanted, because /dev/fd/* itself includes
    /dev again, which lead to unwanted results.
  • user/rpld.c: fixed a filedescriptor leak in fill_info()
  • user/rpld.c: another try at getting ownerships right; rpld still had a case
    where the recorded owner differed from the owner of the /dev node (with
    respect to the design decision). See one below:
  • linux-2.6/rpldev.c: /dev/tty is a special case with respect to ownerships,
    so it must be filtered out in krpl_init() and krpl_open().

doc/rplctl.8.php
<?php include_once("_header.php"); ?>

<h1>rplctl - control rpld</h1>

<p><code>rplctl <b>[</b>-A <i>tty</i><b>] [</b>-D <i>tty</i><b>] [</b>-L
<b>[</b><i>tty</i><b>]] [</b>-S <i>tty</i><b>] [</b>-X <i>tty</i><b>] [</b>-Z
<b>[</b><i>tty</i><b>]] [</b>-f <i>socket</i><b>] [</b>-t<b>]</b></code></p>

<h1>Description</h1>

<p class="block">The <code>rplctl</code> utility is used to query and control
<i>rpld</i> (and its <i>infod</i> subcomponent). It will print info about the
requested ttys, or all ttys currently monitored or optionized if no arguments
are given.</p>

<p class="block"><i>rplctl</i> will read the <code>INFOD_PORT</code> variable
from <code>rpld.conf</code> by default, if not overriden by the <code>-f</code>
option.</p>

<h1>Options</h1>

<table border="1">
  <tr>
    <td class="t1n"><code>-A <i>tty</i></code></td>
    <td class="t1">Activate logging for the given tty.  It can be a filename
      (which must exist) or a device number using
      "<code>=MAJOR,MINOR</code>".</td>
  </tr>
  <tr>
    <td class="t2n"><code>-D <i>tty</i></code></td>
    <td class="t2">Deactivate logging for the given tty.</td>
  </tr>
  <tr>
    <td class="t1n"><code>-L <b>[</b><i>tty</i><b>]</b></td>
    <td class="t1">Explicitly request a listing of all ttys currently active or
      have option set. If a <i>tty</i> is given, only display info for that
      particular one.</td>
  </tr>
  <tr>
    <td class="t2n"><code>-S <i>tty</i></code></td>
    <td class="t2">Deactivate logging for the given tty until it is
      deinitialized.</td>
  </tr>
  <tr>
    <td class="t1n"><code>-X <i>tty</i></code></td>
    <td class="t1">Call <code>log_close()</code> for the given tty. <i>rpld</i>
      will close the logfile, and necessarily open up a new one whenever new
      activity is detected (and logging this terminal is not deactivated) on
      that terminal.</td>
  </tr>
  <tr>
    <td class="t2n"><code>-Z <b>[</b><i>tty</i><b>]</b></code></td>
    <td class="t2">Zero all packet and byte counters, both of <i>rpld</i> and
      every single tty. If a <i>tty</i> is given, only zeroes the stats of that
      tty.</td>
  </tr>
  <tr>
    <td class="t1n"><code>-f <i>socket</i></code></td>
    <td class="t1">Path to the <i>infod</i> socket to connect to. The default
      is <code>/var/run/.rplinfo_socket</code> or <code>INFOD_PORT</code> in
      the <code>rpld.conf</code> configuration file.</td>
  </tr>
  <tr>
    <td class="t2n"><code>-t</td>
    <td class="t2">Generate output suitable for <code>sscanf()</code>,
      <code>split()</code> or anything that deals best with simple text
      strings. See below for details on the output format.</td>
  </tr>
</table>

<h1>Example output</h1>

<p class="block">A command like <code>rplctl</code> could output this:</p>

<p class="code"><code>
A TTY &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;BYTES IN &nbsp; &nbsp; &nbsp; OUT FILENAME<br />
==========================================================================<br />
&nbsp; * &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;2159 &nbsp; &nbsp; 81129 ttyrpld 2.12/Linux<br />
&nbsp; IOCD: 0/118/117/0 &nbsp;RW: 1851/1396 &nbsp;I: 0 &nbsp;B: 0<br />
--------------------------------------------------------------------------<br />
D tty7 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;937 &nbsp; &nbsp; &nbsp; &nbsp; 0 root/20050328.224208.tty7<br />
D pts-1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 0 &nbsp; &nbsp; &nbsp; 586 boinc/20050328.224219.pts-1<br />
D pts-2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 0 &nbsp; &nbsp; &nbsp;5812 root/20050328.224208.pts-2<br />
D pts-3 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 358 &nbsp; &nbsp; 22113 root/20050328.224208.pts-3<br />
D pts-4 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 864 &nbsp; &nbsp; 52618 root/20050328.224341.pts-4<br />
--------------------------------------------------------------------------
</code></p>

<p class="block">The status field (<code>A</code>) can be one of A, D or S:
<b>A</b>ctivated, <b>D</b>eactivated, Deactived for this <b>S</b>ession.</p>

<h1>sscanf() format</h1>

<p class="block">On the other hand, <code>rplctl -t</code> could produce
this:</p>

<p class="code"><code>
ttyrpld 2.12<br />
format 3<br />
0 144 143 0 3347 2496 3756 122881 0 0<br />
D 4 7 tty7 1507 0 /var/log/rpl/root/20050328.224208.tty7<br />
D 136 1 pts-1 0 1241 /var/log/rpl/boinc/20050328.224219.pts-1<br />
D 136 2 pts-2 0 7585 /var/log/rpl/root/20050328.224208.pts-2<br />
D 136 3 pts-3 365 22798 /var/log/rpl/root/20050328.224208.pts-3<br />
D 136 4 pts-4 1884 91257 /var/log/rpl/root/20050328.224341.pts-4<br />
</code></p>

<p class="block">The first line of the "parseable" output (<code>-t</code>) is
the program used. It can (and should) be ignored by programs querying
<i>infod</i>. The second is the status line about <i>rpld</i>. The
<code>sscanf</code> formula is "<code>%lu %lu %lu %lu %llu %llu %lu
%lu</code>". The fields are (in this order):</p>

<ul>
  <li><code>EVT_INIT</code>&nbsp;-- tty first opens</li>
  <li><code>EVT_OPEN</code>&nbsp;-- number of <code>open()</code> operation on
    any tty</li>
  <li><code>EVT_CLOSE</code>&nbsp;-- number of <code>close()</code> operations
    on any tty</li>
  <li><code>EVT_DEINIT</code>&nbsp;-- tty deallocation</li>
  <li><code>EVT_READ</code>&nbsp;-- packets read from <i>rpldev</i></li>
  <li><code>EVT_WRITE</code>&nbsp;-- packets read from <i>rpldev</i></li>
  <li><code>EVT_READ</code>&nbsp;-- payload bytecount</li>
  <li><code>EVT_WRITE</code>&nbsp;-- payload bytecount</li>
  <li><code>EVT_IOCTL</code> packets received</li>
  <li>packets considered bad (i.e. bad magic)&nbsp;-- might be more than there
    are actually damaged, because the algorithm needs to find <i>something</i>
    that looks good.</li>
</ul>

<p class="block">All other lines are tty information lines, whose formula is
"<code>%c %ld %ld %s %llu %llu %s</code>", and the fields are:</p>

<ul>
  <li>status&nbsp;-- <code>'A'</code> for activated, <code>'D'</code> is
    deactivated and <code>'S'</code> means deactivated until session ends.</li>
  <li>major number</li>
  <li>minor number</li>
  <li>mnemonic string for the major/minor number</li>
  <li><code>EVT_READ</code> payload bytecount</li>
  <li><code>EVT_WRITE</code> payload bytecount</li>
  <li>full filename being written to (this is different from the human-readable
    output which only shows it in part)</li>
</ul>

<h1>See also</h1>

<ul>
  <li><a href="rpld.8.php"><code>rpld(8)</code></a></li>
</ul>

<?php include_once("_footer.php"); ?>
